src/txadmission.cpp

```diff
@@ -0,0 +1,1028 @@
+
+            {   // tx must be popped within the TX_PROCESSING corral or the state break between processing
+                // and commitment will not be clean
+                CCriticalBlock lock(csTxInQ, "csTxInQ", __FILE__, __LINE__);
+                if (txInQ.empty()) continue;  // abort back into wait loop if another thread got my tx
+                txd = txInQ.front(); // make copy so I can pop & release
+                txInQ.pop();
+            }
+
+            CTransactionRef &tx = txd.tx;
+            CInv inv(MSG_TX, tx->GetHash());
+
+            std::vector<COutPoint> vCoinsToUncache;
+            {
+                // Check for recently rejected (and do other quick existence checks)
+                bool have = TxAlreadyHave(inv);
+                if (have)
+                {
+                    recentRejects.insert(tx->GetHash());
+
+                    if (txd.whitelisted && GetBoolArg("-whitelistforcerelay", DEFAULT_WHITELISTFORCERELAY))
+                    {
+                        // Always relay transactions received from whitelisted peers, even
+                        // if they were already in the mempool or rejected from it due
+                        // to policy, allowing the node to function as a gateway for
+                        // nodes hidden behind it.
+                        //
+                        // Never relay transactions that we would assign a non-zero DoS
+                        // score for, as we expect peers to do the same with us in that
+                        // case.
+                        int nDoS = 0;
+                        if (!state.IsInvalid(nDoS) || nDoS == 0)
+                        {
+                            LOGA("Force relaying tx %s from whitelisted peer=%s\n", tx->GetHash().ToString(),
+                                txd.nodeName);
+                            RelayTransaction(tx);
+                        }
+                        else
+                        {
+                            LOGA("Not relaying invalid transaction %s from whitelisted peer=%d (%s)\n",
+                                tx->GetHash().ToString(), txd.nodeName, FormatStateMessage(state));
+                        }
+                    }
+                    continue;
+                }
+            }
+
+            bool isRespend = false;
+            if (ParallelAcceptToMemoryPool(
+                    txHandlerSnap, mempool, state, tx, true, &fMissingInputs, false, false, TransactionClass::DEFAULT, vCoinsToUncache, &isRespend))
+            {
+                RelayTransaction(tx);
+
+                //LOG(MEMPOOL, "AcceptToMemoryPool: peer=%s: accepted %s onto Q\n", txd.nodeName,
+                //    tx->GetHash().ToString());
+            }
+            else
+            {
+                for (const COutPoint &remove : vCoinsToUncache)
+                    pcoinsTip->Uncache(remove);
+                if (fMissingInputs)
```

Comment:
-for (const COutPoint &remove : vCoinsToUncache)
-                    pcoinsTip->Uncache(remove);

We keep coins from orphans in the cache...if the orphan is later timed out of the orphan cache we remove them at that time. Reason is because the orphan will in most cases be pulled into the mempool and so we don't have to go and get the coins again, possibly from disk.

Reply:
you're right, this code should be in an else clause of "if (fMissingInputs)" so that we only uncache if the transaction is rejected for another reason.